### ISSUE TITLE
Add support form vim style key bindings

### ISSUE THREAD

**Issue by kylemhall** [darrenburns/posting#34]
It would be really great to have support for vim-style key bindings!

**Comment by darrenburns** [comment]
I assume you mean in the request body editor? 😄

Let me know which keys in particular you would like.

**Comment by kylemhall** [comment]
That is correct!

I grabbed this bit from https://www.freecodecamp.org/news/vim-key-bindings-reference/
I think most of these could be added without too much fuss. I imagine things like insert vs visual mode might be more difficult.

Moving to Top and Bottom

You can use these keys to move to the top or bottom of the page:

gg  --> Move to the first line of the page
G   --> Move to the last line of the page
How to Edit Text in Normal Mode
Deletion and Copying

You can delete or copy characters using these keys:

x   --> Delete the character under the cursor
dd  --> Delete the entire line
yy  --> Copy (yank) the entire line
p   --> Paste the previously deleted or copied text after the cursor

Undo and Redo

u          --> Undo the last action
Ctrl + r   --> Redo the undone action

Thanks!

**Comment by darrenburns** [comment]
These binds would require the addition of an insert mode though. Quite a few Vim bindings already work in the response area since that's read only. However, I would need to add a mode system in order to support the ones you've listed in the request area.

Possibly in the future I'll look at this, but for now you can easily edit the request in Vim by pressing f4 (assuming you have your EDITOR env var set).

**Comment by kylemhall** [comment]
Excellent, thanks for that tip!

**Comment by diegoulloao** [comment]
It would be great to move 10 lines up/down by pressing `ctr+d/ctrl-u` like vim.

**Comment by darrenburns** [comment]
I've added ctrl+d and ctrl+u in 1.2.0.

**Comment by diegoulloao** [comment]
IMHO, due request body is read-only, I think other keybindings for edit shouldn't be included, for editing you can just open the editor. Said that, and the fact that basic movements are already implemented, this issue could be mark as `done`

**Comment by StitiFatah** [comment]
Would love at least basic vim support for the URL bar, 

> t would be great to move 10 lines up/down by pressing `ctr+u/ctrl`

I like the F4 philosophy, vim emulation for editing is always going to be limited and/or would require significant efforts so why not just using (neo)vim ? 

But F4 doesn't seem to work for the URL bar, right ? Would it be possible to enable it there ? It's rather frustrating to have to reach out for the mouse in the terminal.

**Comment by StitiFatah** [comment]
> Would love at least basic vim support for the URL bar,
> 
> > t would be great to move 10 lines up/down by pressing `ctr+u/ctrl`
> 
> I like the F4 philosophy, vim emulation for editing is always going to be limited and/or would require significant efforts so why not just using (neo)vim ?
> 
> But F4 doesn't seem to work for the URL bar, right ? Would it be possible to enable it there ? It's rather frustrating to have to reach out for the mouse in the terminal.



> ITOR env var set).



I like the F4 philosophy, vim emulation for editing is always going to be limited and/or would require significant efforts so why not just using (neo)vim ? 

But F4 doesn't seem to work for the URL bar, right ? Would it be possible to enable it there ? It's rather frustrating to have to reach out for the mouse in the terminal.

**Comment by StitiFatah** [comment]
> These binds would require the addition of an insert mode though. Quite a few Vim bindings already work in the response area since that's read only. However, I would need to add a mode system in order to support the ones you've listed in the request area.
> 
> Possibly in the future I'll look at this, but for now you can easily edit the request in Vim by pressing f4 (assuming you have your EDITOR env var set).

I like the F4 philosophy, vim emulation for editing is always going to be limited and/or would require significant efforts so why not just using (neo)vim ? 

But F4 doesn't seem to work for the URL bar, right ? Would it be possible to enable it there ? It's rather frustrating to have to reach out for the mouse in the terminal.

**Comment by darrenburns** [comment]
Yep, I can add editor support to the URL bar. You're the second person to ask this week and it should be fairly straightforward, so I'll aim to get it in soon.

Out of curiosity though, why do you feel the need to reach for the mouse when using the URL bar? Is there some key binding you would find useful?

**Comment by darrenburns** [comment]
Also, I'm still deciding how best to approach directional navigation. There will likely be something similar to what you suggested, but I'm not 100% sure how it will work yet.

I'm thinking: Ctrl+w followed by lowercase hjkl to move between widgets directionally, or uppercase HJKL to move between "sections" directionally (I.e tabs/collection tree/request area/response area).

I'm also thinking of potentially having a shortcut for this like alt+hjkl and alt+HJKL to do this navigation in a single chord.

**Comment by StitiFatah** [comment]
Thanks for considering it !

Without vim motions I in fact feel the need to use the mouse (or alternatively spam direction keys) to do simple things, if I had to give a restricted number of vim motions which would highly reduce that need I'd say : 

- ^ and $ (or "g_"), to seamlessly go to each side of the URL to edit the domain name or a querystring for example 
-  the y operator and specifically yy to copy the URL to the browser or another client  
-  the d or c operator, specifically dw/cw to change a query parameter
- u/ctrl+r to undo/redo changes

I think those would do the trick for most use cases given the limited scope of a URL bar.

------

For the Ctrl+W concern I think Ctrl+W, even if a bit long to perform, is fine for vim user, the muscle memory and the convenience of not having to memorize new keybindings specific to an application outweigh the inconvenience of its length imo. Unless you plan to add it as a general shortcut for everyone to have a 4 directions and widget/sections aware alternative tab/shift+tab, in that case yeah ctrl+w might be confusing to others I guess.
